### PR TITLE
Deployment name changes

### LINKED ISSUE #1571: Changing names of componentSpecs causes reconcile error
Latest master updates 1.0.3-SNAPSHOT for naming changes for deployments causes a reconcile error on apply updates to a SeldonDeployment that changes its name.

This did not happen before when hash of image was used as part of name.

### PR DESCRIPTION
**Pull request by cliveseldon**
Fixes #1571 

 * Changes the way deployments are named to ensure a new deployment is created if the names of the containers changes. As new labels are created we cannot try to change an existing deployment as label selectors are immutable.
 * Add notebook tests for rolling_update scenarios to check the changes

### CODE REVIEW

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **cliveseldon**

If they are not already assigned, you can assign the PR to them by writing `/assign @cliveseldon` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by seldondev** [comment]
Sun Mar 22 14:51:52 UTC 2020
The logs for [lint] [2] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1573/2.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1573 --build=2

**Comment by seldondev** [comment]
Sun Mar 22 14:51:54 UTC 2020
The logs for [pr-build] [1] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1573/1.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1573 --build=1

File: operator/apis/machinelearning/v1/seldondeployment_types.go [hunk 1]
```diff
@@ -46,6 +46,8 @@ const (
 	ANNOTATION_NO_ENGINE       = "seldon.io/no-engine"
 	ANNOTATION_CUSTOM_SVC_NAME = "seldon.io/svc-name"
 	ANNOTATION_EXECUTOR        = "seldon.io/executor"
+
+	DeploymentNamePredfix = "seldon"
```
Comment (adriangonz) [inline operator/apis/machinelearning/v1/seldondeployment_types.go]: Possible typo in `Predfix`?

File: operator/apis/machinelearning/v1/seldondeployment_types.go [hunk 2]
```diff
@@ -72,15 +74,27 @@ func GetExplainerDeploymentName(sdepName string, predictorSpec *PredictorSpec) s
...
+	}
+	return name
+}
+
+func GetDeploymentName(mlDep *SeldonDeployment, predictorSpec PredictorSpec, podSpec *SeldonPodSpec, podSpecIdx int) string {
+	baseName := mlDep.Name + "-" + predictorSpec.Name + "-" + strconv.Itoa(podSpecIdx) + "-"
+	var name string
 	if podSpec != nil && len(podSpec.Metadata.Name) != 0 {
-		name = name + "-" + podSpec.Metadata.Name
+		name = baseName + podSpec.Metadata.Name
```
Comment (adriangonz) [inline operator/apis/machinelearning/v1/seldondeployment_types.go]: If `podSpec` has a name defined, but the containers change name after an update, wouldn't we run into the same issue we are trying to solve?

Perhaps we could always append the hashed list of container names?

File: operator/controllers/seldondeployment_controller.go [hunk 3]
```diff
@@ -684,6 +684,7 @@ func createContainerService(deploy *appsv1.Deployment, p machinelearningv1.Predi
 
 	//Add labels for this service to deployment
 	deploy.ObjectMeta.Labels[containerServiceKey] = containerServiceValue
+	//FIXME testing this change
```
Comment (adriangonz) [inline operator/controllers/seldondeployment_controller.go]: Is this something we need to fix? :grin:

File: operator/controllers/seldondeployment_controller.go [hunk 4]
```diff
@@ -1411,7 +1413,6 @@ func (r *SeldonDeploymentReconciler) updateStatusForError(desired *machinelearni
 			"Failed to update status for SeldonDeployment %q: %v", desired.Name, err)
 	} else {
```
Comment (adriangonz) [inline operator/controllers/seldondeployment_controller.go]: Can we remove the `else` if there isn't anything there anymore?

File: operator/controllers/seldondeployment_controller.go [hunk 5]
```diff
@@ -1430,7 +1431,6 @@ func (r *SeldonDeploymentReconciler) updateStatus(desired *machinelearningv1.Sel
 		return err
 	} else {
 		// If there was a difference and there was no error.
-		r.Recorder.Eventf(desired, corev1.EventTypeNormal, constants.EventsUpdated, "Updated SeldonDeployment %q", desired.GetName())
```
Comment (adriangonz) [inline operator/controllers/seldondeployment_controller.go]: Same as above, can we remove the `else` if there isn't anything there anymore?

File: operator/apis/machinelearning/v1/seldondeployment_types.go (hunk 2 above)
Comment (cliveseldon) [reply operator/apis/machinelearning/v1/seldondeployment_types.go]: If you specify the name in the podSpec meta then you are taking control of whether its a new or updated deployment that is created.

**Comment by seldondev** [comment]
Wed Mar 25 09:56:19 UTC 2020
The logs for [pr-build] [3] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1573/3.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1573 --build=3

**Comment by seldondev** [comment]
Wed Mar 25 09:56:19 UTC 2020
The logs for [lint] [4] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1573/4.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1573 --build=4

**Comment by seldondev** [comment]
Wed Mar 25 09:56:29 UTC 2020
The logs for [pr-build] [5] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1573/5.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1573 --build=5

**Comment by seldondev** [comment]
Wed Mar 25 09:56:46 UTC 2020
The logs for [lint] [6] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1573/6.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1573 --build=6

**Comment by seldondev** [comment]
Wed Mar 25 10:22:59 UTC 2020
The logs for [integration] [7] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1573/7.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1573 --build=7

**Comment by seldondev** [comment]
@cliveseldon: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | a51f0bece63aa4d9455fafa802120b4de14ad020 | [link]() | `/test integration`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/SeldonIO/seldon-core/pull/1573#issuecomment-604306234" title="Approved">adriangonz</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=SeldonIO%2Fseldon-core).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/SeldonIO/seldon-core/blob/master/OWNERS)~~ [adriangonz]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by seldondev** [comment]
Thu Mar 26 10:28:22 UTC 2020
The logs for [lint] [9] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1573/9.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1573 --build=9

**Comment by seldondev** [comment]
Thu Mar 26 10:28:25 UTC 2020
The logs for [pr-build] [8] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1573/8.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1573 --build=8

**Comment by seldondev** [comment]
@cliveseldon: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | a51f0bece63aa4d9455fafa802120b4de14ad020 | [link]() | `/test integration`
lint | 49bd6818d259d293a11eccb924fe2529b7d3004a | [link]() | `/test lint`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@cliveseldon: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | a51f0bece63aa4d9455fafa802120b4de14ad020 | [link]() | `/test integration`
lint | 49bd6818d259d293a11eccb924fe2529b7d3004a | [link]() | `/test lint`
pr-build | 49bd6818d259d293a11eccb924fe2529b7d3004a | [link]() | `/test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@cliveseldon: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | a51f0bece63aa4d9455fafa802120b4de14ad020 | [link]() | `/test integration`
lint | 49bd6818d259d293a11eccb924fe2529b7d3004a | [link]() | `/test lint`
pr-build | 49bd6818d259d293a11eccb924fe2529b7d3004a | [link]() | `/test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
@cliveseldon: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
integration | a51f0bece63aa4d9455fafa802120b4de14ad020 | [link]() | `/test integration`
pr-build | 49bd6818d259d293a11eccb924fe2529b7d3004a | [link]() | `/test this`
lint | 49bd6818d259d293a11eccb924fe2529b7d3004a | [link]() | `/test lint`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by seldondev** [comment]
Thu Mar 26 20:51:48 UTC 2020
The logs for [lint] [15] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1573/15.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1573 --build=15

**Comment by seldondev** [comment]
Thu Mar 26 20:51:51 UTC 2020
The logs for [pr-build] [14] will show after the pipeline context has finished.  
https://github.com/SeldonIO/seldon-core/blob/gh-pages/jenkins-x/logs/SeldonIO/seldon-core/PR-1573/14.log

impatient try  
jx get build logs SeldonIO/seldon-core/PR-1573 --build=14

**Pull request merged**